components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -2038,47 +2029,66 @@ impl Animate for MatrixDecomposed3D {
-            let b = &other.quaternion;
-            sum.quaternion = Quaternion(
+            let a = &scaled_from;
+            let b = to;
+            return Ok(Quaternion(
                 a.3 * b.0 + a.0 * b.3 + a.1 * b.2 - a.2 * b.1,
                 a.3 * b.1 - a.0 * b.2 + a.1 * b.3 + a.2 * b.0,
                 a.3 * b.2 + a.0 * b.1 - a.1 * b.0 + a.2 * b.3,
                 a.3 * b.3 - a.0 * b.0 - a.1 * b.1 - a.2 * b.2,
-            );
-        } else {
-            let mut product = self.quaternion.0 * other.quaternion.0 +
-                              self.quaternion.1 * other.quaternion.1 +
-                              self.quaternion.2 * other.quaternion.2 +
-                              self.quaternion.3 * other.quaternion.3;
-
-            // Clamp product to -1.0 <= product <= 1.0
-            product = product.min(1.0);
-            product = product.max(-1.0);
+            ));
+        }
 
-            if product == 1.0 {
-                return Ok(sum);
-            }
+        let mut product = from.0 * to.0 +
+                          from.1 * to.1 +
+                          from.2 * to.2 +
+                          from.3 * to.3;
 
-            let theta = product.acos();
-            let w = (other_weight * theta).sin() * 1.0 / (1.0 - product * product).sqrt();
+        // Clamp product to -1.0 <= product <= 1.0
+        product = product.min(1.0);
+        product = product.max(-1.0);
 
-            let mut a = *self;
-            let mut b = *other;
-            % for i in range(4):
-                a.quaternion.${i} *= (other_weight * theta).cos() - product * w;
-                b.quaternion.${i} *= w;
-                sum.quaternion.${i} = a.quaternion.${i} + b.quaternion.${i};
-            % endfor
+        if product == 1.0 {
+            return Ok(*from);
         }
 
+        let theta = product.acos();
+        let w = (other_weight * theta).sin() * 1.0 / (1.0 - product * product).sqrt();
+
+        let mut a = *from;
+        let mut b = *to;
+        let mut result = Quaternion(0., 0., 0., 0.,);
+        % for i in range(4):
+            a.${i} *= (other_weight * theta).cos() - product * w;
+            b.${i} *= w;
+            result.${i} = a.${i} + b.${i};
+        % endfor
+
+        Ok(result)
+    }
+}
+
+impl Animate for MatrixDecomposed3D {
+    /// <https://drafts.csswg.org/css-transforms/#interpolation-of-decomposed-3d-matrix-values>
+    fn animate(&self, other: &Self, procedure: Procedure) -> Result<Self, ()> {
+        let mut sum = *self;
+
+        // Add translate, scale, skew and perspective components.
+        sum.translate = self.translate.animate(&other.translate, procedure)?;
+        sum.scale = self.scale.animate(&other.scale, procedure)?;
+        sum.skew = self.skew.animate(&other.skew, procedure)?;
+        sum.perspective = self.perspective.animate(&other.perspective, procedure)?;
+        sum.quaternion = MatrixDecomposed3D::slerp(&self.quaternion,
```

Comment:
Why is this call not like the others? Can't the type of `self.quaternion` implement `Animate` and use `MatrixDecomposed3D::slerp` there? That would make this implementation for `MatrixDecomposed3D` derivable.